### PR TITLE
Fix PR push times

### PR DESCRIPTION
**Pull request by mark-i-m**
Currently if you push on a different branch on your fork, it will reset the time for all PRs from that fork, even if they are completely unrelated. This issue aims to fix that.

### CODE REVIEW

**Review by PlasmaPower — APPROVED** [review]

**Review by andrewda — APPROVED** [review]

**Comment by PlasmaPower** [comment]
@mark-i-m Please fix the tests.

**Comment by mark-i-m** [comment]
Working on it :)

**Comment by PlasmaPower** [comment]
Still not fixed :(

If you could also expand the tests to with and without events present, that'd be great.

**Comment by PlasmaPower** [comment]
and I'm assuming you know this, but you can run the tests locally of course.

**Comment by mark-i-m** [comment]
Hmm... I am actually having trouble running them locally... which is why I am doing a lot of guessing...

**Comment by mark-i-m** [comment]
What things do I need to pip install to run them locally?

**Comment by andrewda** [comment]
Now it's just a linting issue 😝

File: tests/github_api/prs_test.py [hunk 1]
```diff
@@ -12,8 +12,13 @@ def create_mock_pr(number, title, pushed_at):
         "statuses_url": "statuses_url/{}".format(number),
         "head": {
             "repo": {
-                "pushed_at": pushed_at
-            }
+                "pushed_at": pushed_at,
+                "name": "test_repo",
+            },
+            "ref" : "test_ref"
```
Comment (andrewda) [inline tests/github_api/prs_test.py]: Remove space after `"ref"`

**Review by andrewda — CHANGES_REQUESTED** [review]

**Comment by mark-i-m** [comment]
> If you could also expand the tests to with and without events present, that'd be great.

First, I need to figure out how the testing harness works... :stuck_out_tongue:

**Comment by PlasmaPower** [comment]
@mark-i-m 

```
pip install -r requirements.txt
python -m unittest
```

`pip` might need root/admin, depending on your setup.

**Review by PlasmaPower — APPROVED** [review]
I would still appreciate tests for the event driven code, but I approve this PR as it stands

**Comment by mark-i-m** [comment]
```$ python -m unittest github_api
Traceback (most recent call last):
  File "/usr/lib/python2.7/runpy.py", line 174, in _run_module_as_main
    "__main__", fname, loader, pkg_name)
  File "/usr/lib/python2.7/runpy.py", line 72, in _run_code
    exec code in run_globals
  File "/usr/lib/python2.7/unittest/__main__.py", line 12, in <module>
    main(module=None)
  File "/usr/lib/python2.7/unittest/main.py", line 94, in __init__
    self.parseArgs(argv)
  File "/usr/lib/python2.7/unittest/main.py", line 149, in parseArgs
    self.createTests()
  File "/usr/lib/python2.7/unittest/main.py", line 158, in createTests
    self.module)
  File "/usr/lib/python2.7/unittest/loader.py", line 130, in loadTestsFromNames
    suites = [self.loadTestsFromName(name, module) for name in names]
  File "/usr/lib/python2.7/unittest/loader.py", line 91, in loadTestsFromName
    module = __import__('.'.join(parts_copy))
  File "github_api/__init__.py", line 7, in <module>
    import settings
  File "settings.py", line 2, in <module>
    import misc
  File "misc.py", line 2, in <module>
    from urllib.parse import urlparse
ImportError: No module named parse```

**Review by andrewda — APPROVED** [review]
Also would like some tests if possible, but this looks good 👍

**Comment by PlasmaPower** [comment]
@mark-i-m Python 3.6? If so, our `requirements.txt` might be off. I'll look into it.

**Comment by mark-i-m** [comment]
Ah, that's true... ubuntu 16.04 only has python 3.5.2... Is there any way to run this in the docker container?

**Comment by PlasmaPower** [comment]
@mark-i-m No idea, and that Python version might not even be a problem, IDK.

**Comment by andrewda** [comment]
Tests pass, just linting now 👍

File: tests/github_api/prs_test.py [hunk 2]
```diff
@@ -103,3 +124,33 @@ def get_is_mergeable_side_effect(api, urn, pr_num):
...
+            create_mock_events([("PushEvent", "test_ref", "2017-01-01T00:00:10Z"),
+                                ("PushEvent", "blah", "2017-01-03T00:00:10Z")])
+
+        api = MagicMock()
+        api.BASE_URL = "api_base_url"
+
+        last_updated = prs.get_pr_last_updated(api,
+                create_mock_pr(10, "OK", "2017-01-02T00:00:00Z"))
+
+        self.assertEquals(last_updated, arrow.get("2017-01-01T00:00:10Z"))
```
Comment (andrewda) [inline tests/github_api/prs_test.py]: `assertEquals` is deprecated in favor of `assertEqual` – small change but would get rid of the warning in the tests. Also on line 156

**Review by andrewda — CHANGES_REQUESTED** [review]

**Comment by PlasmaPower** [comment]
Just a couple of indentation problems now, then it looks good to go!

**Comment by PlasmaPower** [comment]
I'm assuming you have flake8 working locally?

**Comment by mark-i-m** [comment]
TBH, I was playing around with the Dockerfile trying to run things in their... that seems like a much preferable option to installing things locally.

File: github_api/prs.py [hunk 3]
```diff
@@ -122,14 +122,48 @@ def close_pr(api, urn, pr):
...
+    pr_ref = pr_data["head"]["ref"]
+    pr_repo = pr_data["head"]["repo"]["name"]
+    pr_owner = pr_data["user"]["login"]
+
+    events = get_events(api, pr_owner, pr_repo)
+    events = list(filter(lambda e: e["type"] == "PushEvent", events))
+
+    # Gives the full ref name "ref/heads/my_branch_name", but we just
+    # want my_branch_name, so isolate it...
+    events = list(filter(lambda e: e["payload"]["ref"].split("/")[-1] == pr_ref, events))
```
Comment (PlasmaPower) [inline github_api/prs.py]: Branch names can contain `/`s IIRC - try `'/'.join(arr[2:])` or similar.

**Review by PlasmaPower — CHANGES_REQUESTED** [review]

**Comment by mark-i-m** [comment]
Yay :tada: It works. Shall I unWIP?

**Comment by PlasmaPower** [comment]
@mark-i-m almost, but I just noticed something. See above comment.

File: github_api/prs.py [hunk 4]
```diff
@@ -122,14 +122,48 @@ def close_pr(api, urn, pr):
...
-        return arrow.get(repo["pushed_at"])
-    else:
-        return None
+
+    pr_ref = pr_data["head"]["ref"]
+    pr_repo = pr_data["head"]["repo"]["name"]
+    pr_owner = pr_data["user"]["login"]
+
+    events = get_events(api, pr_owner, pr_repo)
+    events = list(filter(lambda e: e["type"] == "PushEvent", events))
```
Comment (PlasmaPower) [inline github_api/prs.py]: Do you need to turn it into a list here? I think `filter` can take an iterator.

**Comment by PlasmaPower** [comment]
One other thing I forgot: you should probably call `max(current_return_value, pr["created_at"])`.

File: github_api/prs.py (hunk 4 above)
Comment (mark-i-m) [reply github_api/prs.py]: Ah, yeah, I was printing the value so I need it as a list. Will fix

**Comment by mark-i-m** [comment]
@PlasmaPower why?

**Comment by PlasmaPower** [comment]
@mark-i-m mark-i-m A branch can be pushed to, then later the PR created for it. We don't want an insta-close possible in that situation. It's happened before.

**Comment by mark-i-m** [comment]
Hmm... that seems like it should be caller's problem, though?

**Comment by PlasmaPower** [comment]
@mark-i-m sorta, but we should still handle it since it's easily possible.

**Comment by mark-i-m** [comment]
The problem is that `max(current_return_value, pr["created_at"])` returns `current_return_value` always, right?

File: github_api/prs.py [hunk 5]
```diff
@@ -147,11 +147,11 @@ def get_pr_last_updated(api, pr_data):
     pr_owner = pr_data["user"]["login"]
 
     events = get_events(api, pr_owner, pr_repo)
-    events = list(filter(lambda e: e["type"] == "PushEvent", events))
+    events = filter(lambda e: e["type"] == "PushEvent", events)
 
     # Gives the full ref name "ref/heads/my_branch_name", but we just
     # want my_branch_name, so isolate it...
-    ref_name = lambda e: '/'.join(e["payload"]["ref"].split("/")[2:]) == pr_ref
+    ref_name = lambda e: '/'.join(e["payload"]["ref"].split("/")[3:]) == pr_ref
```
Comment (PlasmaPower) [inline github_api/prs.py]: You sure about `[3:]`? wouldn't `my_branch_name` be index 2?

**Comment by PlasmaPower** [comment]
@mark-i-m not always - reread my example:

1. Author pushes to `example-branch`
2. Author waits a couple hours
3. Author makes a PR for `example-branch`
4. Author's PR gets instantly closed
    - Current return value: time of 1 (too soon)
    - `pr["created_at"]`: time of 3 (correct)

This can also happen when a PR is recreated to address conflicts/other issues. It's happened before.

File: github_api/prs.py (hunk 5 above)
Comment (mark-i-m) [reply github_api/prs.py]: "/ref/head/blah" => ["", "ref", "head", "blah"]

**Comment by mark-i-m** [comment]
Oh, I thought by `current_return_value` you meant the value without this PR... I see now that I misread your post. Yes, that makes sense.

**Comment by PlasmaPower** [comment]
Oh I could see how that would be confusing. Great work on this! It's a shame it's moving forward the voting window for your other PRs haha.

**Comment by mark-i-m** [comment]
Lol, the irony :laughing:

**Comment by mark-i-m** [comment]
Sorry, everyone... I need to sleep... will finish this up tomorrow :zzz:

**Comment by PlasmaPower** [comment]
Sounds good! Can't wait.

File: github_api/prs.py [hunk 6]
```diff
@@ -122,14 +122,52 @@ def close_pr(api, urn, pr):
...
+    def ref_name(e):
+        '/'.join(e["payload"]["ref"].split("/")[3:]) == pr_ref
+
+    events = list(filter(ref_name, events))
+
+    if len(events) == 0:
+        # if we can't get good data, fall back to repo push time
+        repo = pr_data["head"]["repo"]
+        if repo:
+            return arrow.get(repo["pushed_at"])
```
Comment (PlasmaPower) [inline github_api/prs.py]: I think we want to max this with `pr_data["created_at"]` too.

**Comment by PlasmaPower** [comment]
> I need to sleep... will finish this up tomorrow

Sleep... nah

**Comment by mdcfe** [comment]
/essay
😛

**Comment by andrewda** [comment]
wow @anythingbot good read 👍  (guess I should probably sleep now 💤 )

**Comment by mdcfe** [comment]
~~Meanwhile, in the UK...~~

File: github_api/prs.py (hunk 6 above)
Comment (mark-i-m) [reply github_api/prs.py]: Hmmm... then how did it work before?

Comment (mark-i-m) [reply github_api/prs.py]: I mean, it makes sense

Comment (PlasmaPower) [reply github_api/prs.py]: It didn't work. A number of PRs got insta-closed and one got insta-merged before the threshold increase.

**Comment by mark-i-m** [comment]
Can anyone decipher the error on that last failing test?

**Comment by mark-i-m** [comment]
Nvm... got it :tada:

**Comment by PlasmaPower** [comment]
@mark-i-m I realized the patches were getting mixed up - never would have guess it needs to be in reverse order.

**Comment by mark-i-m** [comment]
Yeah, that's pretty unintuitive

**Comment by PlasmaPower** [comment]
Looks like that arg line is a bit too long.

**Comment by PlasmaPower** [comment]
@mark-i-m Wrong line - it's the arg line of the test we were working on.

**Review by PlasmaPower — APPROVED** [review]
Looks good! :tada:

File: github_api/prs.py (hunk 5 above)
Comment (PlasmaPower) [reply github_api/prs.py]: Oh, that makes a lot of sense. Thanks!

**Pull request merged**